### ISSUE TITLE
`set -gx SHELL fish` causes ssh to fail on mac when tunneling through a bastion server

### ISSUE THREAD

**Issue by b33j0r** [fish-shell/fish-shell#4745]
```sh
~> ssh notafakedomain.cat
fish: No such file or directory
ssh_exchange_identification: Connection closed by remote host
```

# Background

**tl;dr:** *looks like some very masked problem with opening non-interactive subshells from commands such as `ssh -W`, when the startup of that shell misbehaves.*

I devops for a client (mostly a python engineer/architect in other contexts, long story). The infrastructure uses a classic ssh bastion/jump server topology. Important because this issue only breaks my workflow when tunneling ssh via `ProxyCommand` or `ProxyJump`.

It's starting a new shell under the hood, and that's failing with ungooglable error messages. Oh, and preventing connections to production servers! Almost forgot 😛

The only thing I have found that has changed since it worked very well--which is pretty damn funny to me in retrospect--I added `set SHELL` to my `config.fish`

Judging from previous conversations I was able to find, the resolution is likely in better education for users + upstream + downstream (#2160, #2405). I tried to SEO the crap out of my title to help my fellow seekers, apologies if that confuses anyone.

## Environment

Two macs, unfortunately with very much identical configurations due to my `dotfiles` being a bit conformist.

* macOS High Sierra 10.13.3
* fish 2.7.1 (homebrew install)
* had omf installed

OMF seems unrelated. I uninstalled, reinstalled, etc, and ran the `home=tempdir` experiment. But you'll see that I believe this to be upstream to a certain extent, so it was worthy of mention.

### `~/.ssh/config`:

Please accept this as a working ssh config:

```
Host bastion
  HostName bastion.cats.blog.awesome
  # other things here, identity, agent forwarding, muxing, etc

Host service
  HostName service.cats.blog.awesome
  ProxyJump bastion
```

As proof that it doesn't matter to the report, I was always able to do this:

```
~> ssh bastion
# HI WELCOME TO OUR GLORIOUS BASTION SERVER
b33j0r@bastion$ ssh service
# NO PROBLEM, BRO!
b33j0r@service ~>
```

### `~/.config/fish/config.fish`

```
set -gx PATH $HOME/bin $HOME/.pyenv/bin $HOME/.pyenv/shims /usr/local/bin /usr/local/sbin /usr/bin /bin /usr/sbin /sbin
set -gx PYENV_ROOT $HOME/.pyenv
set -gx EDITOR vim
set -gx SHELL fish

status --is-interactive; and source (pyenv init -|psub)

test -e {$HOME}/.iterm2_shell_integration.fish ; and source {$HOME}/.iterm2_shell_integration.fish
```

# Reproduce

I don't know how to simulate this exact situation without having the actual network hosts... off of the top of my head. But if the above makes sense, here's what happens:

## Failing Command

```sh
~> ssh service
fish: No such file or directory
ssh_exchange_identification: Connection closed by remote host
```

Imagine my surprise at seeing `fish` in that output! Makes more sense now, but it was weird for a long time. Led to a bunch of red-herrings, but that's debugging. Verbose with redacted paths:

```sh
OpenSSH_7.6p1, LibreSSL 2.6.2
debug1: Reading configuration data ~/.ssh/config
debug1: ~/.ssh/config line 1: Applying options for *
debug1: ~/.ssh/config line 53: Applying options for service
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 48: Applying options for *
debug1: Executing proxy command: exec ssh -q -W service bastion
debug1: identity file ~/.ssh/id_rsa type 0
debug1: key_load_public: No such file or directory
debug1: identity file ~/.ssh/id_rsa-cert type -1
debug1: key_load_public: No such file or directory
debug1: identity file service.pem type -1
debug1: key_load_public: No such file or directory
debug1: identity file service.pem-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_7.6
debug1: permanently_drop_suid: 501
fish: No such file or directory
ssh_exchange_identification: Connection closed by remote host
```

### Very Important Clues

Why is `fish` in this output? I'm obviously not even reaching the login shell. And in this case, `fish` isn't installed on `bastion` (some things are perplexingly out of my control, folks!) So I did this next, on my local host:

```sh
~> bash
bash-3.2$ ssh service
fish: No such file or directory
ssh_exchange_identification: Connection closed by remote host
```

We start to outgrow my knowledge of shell and subshell implementation details, but I at least knew to try changing the login shell (although ignorantly, first on the remote hosts lol). This works on my local host without other changes:

```
chsh -s (which bash)
```

`<new session>`

```
ugly-bash-prompt$ ssh service
# NO PROBLEM, BRO!
b33j0r@service ~> 
```

So it started to dawn on me that it had something to do with the startup of the subshell, and probably stdin/stdout/stderr (or terminal emulation things I don't remember from college). But the exact error was being masked, and I realized that I was out of my league on an analytical solution.

So I started messing with my fish config to eliminate variables. After a lot of unproductive experiments, I found the "culprit":

### Startling Workaround

```sh
## `config.fish`

set -gx PATH $HOME/bin $HOME/.pyenv/bin $HOME/.pyenv/shims /usr/local/bin /usr/local/sbin /usr/bin /bin /usr/sbin /sbin
set -gx PYENV_ROOT $HOME/.pyenv
set -gx EDITOR vim

## this [indirectly] breaks ssh ProxyJump/ProxyCommand
# set -gx SHELL fish

## this last part was suspect after I learned of the nature of the problem,
## but turns out to not matter. can be commented, deleted, unguarded...
## no effect on any of the behaviors described in this issue

if status --is-interactive

  source (pyenv init -|psub)
  test -e {$HOME}/.iterm2_shell_integration.fish ; and source {$HOME}/.iterm2_shell_integration.fish

end
```

### Almost Definitely Related

I'm very certain that it's not an "ssh thing," but that's how it finally came up for me:

* #2160
* #2405

It's really hard to google, and honestly, requires a lot of experience to debug. I'm a senior-level guy in most of the related areas, and I didn't know what the hell questions to ask for days 😉 To quote the last mention I found:

> [I feel like we've had this question a few times, although I can't see any direct duplicates - it might be worth adding to the FAQ if I'm not making that up.](https://github.com/fish-shell/fish-shell/issues/2405#issuecomment-286982566)

To top it off, a bitch to reproduce... when you're talking about bastion services and VPCs, everything has to be redacted to comfortably share enough information. So if you've read this far, I respect you!

Finally, if it wasn't clear from above, I don't really mean that `set -gx` or the `SHELL` variable having a value is broken in `fish`. I have just reached the point where:

1. There have gotta be domain experts who know how to fill in some of these blanks
2. Create new and more recent searchable results for the next poor souls

If I only accomplish one of those goals, it will have been worth the time.

PS: to make my point a little more forcibly about the difficulty of using search engines to find this topic... c'mon, who hasn't tried googling `terminal fish shell` or `shell fish change 256 color` ... imagine trying to search for `fish shell set SHELL fish` 😆

**Comment by b33j0r** [comment]
Thinking about my report, one more addendum. I don't want to downplay my continued confusion about the original exception:

```
fish: No such file or directory
```

I still don't understand the mechanism that causes the definition of `$SHELL` to lead to an IO error... That's definitely one of the places where I hope the community has implementation insight that I lack. I absolutely thrive in environments where someone points out that I forgot a punctuation mark somewhere

**Comment by faho** [comment]
>fish: No such file or directory

I'm not actually sure this message is printed _by fish_. Because, as you say:

>And in this case, fish isn't installed on bastion

So, we know that:

- ssh starts a non-interactive shell as part of its initialization (which is why #2160 and #2405 happen), at just the time you see this error here.

- fish isn't installed on that server.

So I would assume that what happens:

- You set $SHELL to fish, which is sent to the server

- That server then tries to start that $SHELL to exchange keys

- It doesn't find it and gives up

To confirm this, set SHELL to e.g. zsh (or some other shell that isn't installed on that server). You should see an error like "zsh: No such file or directory" instead.

See also [the openssh documentation for ProxyCommand](https://man.openbsd.org/ssh_config#ProxyCommand):

>The command string extends to the end of the line, and is executed using the user's shell ‘exec’ directive to avoid a lingering shell process.

**Comment by zanchey** [comment]
It's a local problem.

```
> tail -n6 .ssh/config
Host bastion
   Hostname localhost

Host service
   Hostname localhost
   ProxyJump bastion
> env SHELL=fish ssh service
fish: No such file or directory
ssh_exchange_identification: Connection closed by remote host
> env SHELL=asdffish ssh service
asdffish: No such file or directory
ssh_exchange_identification: Connection closed by remote host
> env SHELL=/usr/local/bin/fish ssh service
Password:
Password:
Last login: Sat Feb 24 17:52:28 2018 from ::1
```

On Linux, `SHELL=fish` works, but `SHELL=asdffish` doesn't (with a slightly different error: `write: Broken pipe`).

Try setting your shell to a full path, as apparently SSH doesn't search your $PATH.

**Comment by zanchey** [comment]
`ssh` doesn't send the SHELL environment variable over the connection, unless specified otherwise by the SendEnv directive, so I don't think it's a bastion host problem.

**Comment by b33j0r** [comment]
@zanchey confirmed for me! it appears to be `ssh` handling of `PATH` as stated. With the additional information that `ssh` doesn't send `SHELL`, I believe this to be a complete practical solution. I really hope that others in a similar situation are able to find this post.

I feel like I _should_, at this point, understand why it only happens when `exec` spawns a shell to run a proxy command, but I'll admit that I still have gaps. I'll leave this open for the day in case anyone wants to add to that, otherwise I'll go ahead and close this. Thanks!

**Comment by martonbalazs** [comment]
Hi, I've run into this (or something very similar) on a clean fresh Arch+Xfce4 install. I've set my default shell to Fish, and ssh ProxyCommand doesn't work. It gets stuck at a line

`debug1: ssh_exchange_identification: \033[1 qSSH-2.0-OpenSSH_7.4`

It happens to Xfce4-terminal, Gnome-terminal, xterm under Xfce4. However, it does NOT happen if I go out of Xfce4 (CTRL-ALT-F2) while still running Fish. Resetting my default shell to Bash, the problem goes away under Xfce4 as well. The remote server does not have fish installed.

My config.fish is almost empty, just has

`fish_vi_key_bindings`

and a few innocent aliases. My SHELL variable is /usr/bin/fish.

A very similar setup (with Fish and ssh) on an older Ubuntu 16.04 LTS does not show any problems. Please let me know if there's any test I can do to help.

**Comment by martonbalazs** [comment]
I was playing with the config, weeding out `fish_vi_key_bindings` and noticed that previously my fishd._hostname_ had this line:

`SET fish_key_bindings:fish_vi_key_bindings`

which now disappeared and: problem gone. If I re-insert it in this file manually then ssh proxy breaks again. Luckily, this line does not seem to come back on its own even though I re-introduced `fish_vi_key_bindings` in the config. I have no idea how it appeared in the first place.

**Comment by zanchey** [comment]
You should make sure your line `fish_vi_key_bindings` is protected by an `if status is-interactive` block - otherwise it will output some sequences which the SSH process is not expecting. Like this:

```fish
if status is-interactive
    fish_vi_key_bindings
end
```

**Comment by martonbalazs** [comment]
Thanks, I'll do that and check if the problem returns.

**Comment by faho** [comment]
Urgh...

>`\033[1 q`

That's the "change cursor shape" sequence. We send it because vi-bindings are enabled, and we determined that your terminal can handle it, so we change the cursor shape depending on the vi-mode you are in.

I'm going to make it so we skip that when not interactive, because this is rather annoying - setting the bind mode most of the time prints nothing. Only in this specific case (which depends on the terminal used, so it's surprising and hard to reproduce) does it print anything.

**Comment by tebriel** [comment]
I had a similar problem, my `shell` directive in `.screenrc` was set to `shell "fish"` instead of `shell "/usr/bin/fish"`. Screen would always launch fish shells (likely searching path) but ssh would fail.

**Comment by pickfire** [comment]
Ah, so the solution for this is rather than `export SHELL=fish` we do `export SHELL=/usr/local/bin/fish`, seemed weird to me though.
